### PR TITLE
[WIP] rust: update to 1.32.0.

### PR DESCRIPTION
**Pull request by Cogitri**
@q66

### CODE REVIEW

**Comment by Cogitri** [comment]
Please rebuild for x86_64-musl @Gottox

File: srcpkgs/rust/template [hunk 1]
```diff
@@ -58,10 +58,21 @@ else
...
-		 14de38321f6f2c64496e77bac793ad5511d37696b9c33283bb6f395dc590c4d9
+		 45e633c985c259a13d3b096662241bd8255a86ed4917e0fec396e7071c04c734
+		 69bd6064f605c1d0d7d41070efea1e329d75755b04fbc5a13ffc53efdc3a97c7
 		 200cf76636ccbb153d5c430a5c0f3ad5af206899f76470e3cd046679a5a4ab3f"
 		;;
+	ppc64le)
+		distfiles+="
+		 https://static.rust-lang.org/dist/rustc-${_rust_dist_version}-powerpc64le-unknown-linux-gnu.tar.xz
+		 https://static.rust-lang.org/dist/rustc-${_rust_dist_version}-powerpc64le-unknown-linux-gnu.tar.xz
+		 https://void-ppc64.octaforge.org/distfiles/cargo-${_cargo_dist_version}-powerpc64le-unknown-linux-gnu.tar.xz"
```
Comment (Cogitri) [inline srcpkgs/rust/template]: Should this be mirrored on Void infra @Gottox?

Comment (q66) [reply srcpkgs/rust/template]: I believe all should be mirrored. I mean, I can host my distfiles forever, but stuff like this should probably be mirrored.

**Comment by q66** [comment]
Getting everything built so we can have bootstrap from all ppc64 targets.

Note on cargo from void-ppc64: The official cargo build is broken (`SIGILL`), so sadly it cannot be used. The compiler and standard library are fine. The bug is reported upstream, but nobody has acted on it yet. It's actually not even a bug in cargo itself, more like, misconfiguration on their build server.

**Comment by Cogitri** [comment]
Tested&Built on:

- [ ] armv6l
- [ ] armv6l-musl
- [ ] armv7l
- [ ] armv7l-musl
- [ ] aarch64
- [ ] aarch64-musl
- [ ] i686
- [ ] x86_64
- [ ] x86_64-musl

**Comment by q66** [comment]
We have a cross-compiling problem. Rust tries to build `lzma-sys` for the host system during its build, which uses autotools to configure itself. Subsequently, it fails, because the cross env vars leak in (particularly `LDFLAGS`, `CPP`, `LD`, `NM`, `OBJDUMP`, `STRIP`).

**Comment by q66** [comment]
I tried a workaround for this, by changing `do_build()` to something like

```
do_build() {
	env CFLAGS_${_build_triplet}="${CFLAGS_host}" \
		CPP="${CPP_host}" LD="${LD_host}" NM="${NM_host}" \
		OBJDUMP="${OBJDUMP_host}" STRIP="${STRIP_host}" \
		LDFLAGS="${LDFLAGS_host}" \
		make ${makejobs} ${make_build_args}
}
```

That worked and got me a working rustc for target, but is hardly ideal. Obviously this also does not work when *not* cross compiling.

**Comment by Cogitri** [comment]
Yeah I think either jnbr or me will get to fix it in lzma-sys tomorrow

**Comment by Gottox** [comment]
building.

**Comment by Cogitri** [comment]
Updated the checksums.

**Comment by q66** [comment]
I have all the snapshot builds ready for mirroring (`rustc`+`rust-std` for `ppc64le-musl`/`ppc64-musl` and `cargo` for `ppc64le`/`ppc64le-musl`/`ppc64-musl`). You can find them in https://void-ppc64.octaforge.org/distfiles. 
There is also a `CHECKSUMS.txt` file so you can verify them.

@Gottox @Cogitri

**Comment by Cogitri** [comment]
Nice. Will test armv* now. Did you also test x86_64 -> ppc64le cross, @q66?

**Comment by q66** [comment]
No. I don't have an x86_64 machine ready. I don't see why it would fail though, it's a crossbuild like any other.

**Comment by Cogitri** [comment]
Alright, I'll test cross building to ppc64le then, better to be safe than sorry, eh?

**Comment by q66** [comment]
Yeah, sure. I don't expect you to run into problems, though :) But it will have to compile llvm and a bunch of other stuff too, since we don't yet have things in the repository.

**Comment by Cogitri** [comment]
As mentioned in IRC, FF 64.x compilation fails with Rust 1.32:

https://gist.githubusercontent.com/q66/6b5586383da0d661052b2e6eb293c514/raw/5633c8e6baa4d6a40804fa76f5fbcdce45379f72/gistfile1.txt

https://gist.github.com/q66/75bc033658d2fa20c005309a99c7fdba

**Comment by q66** [comment]
I think it has to be a combination of 1.32 and LLVM 7, and possibly also my powerpc64le target. It will need testing whether it also fails like that on x86_64. I did successfully build FF 64 with 1.31 before that (not sure if it was LLVM 7 or LLVM 6 though).

**Comment by Cogitri** [comment]
I can try building FF on x86_64.

**Comment by q66** [comment]
I updated the template to use bundled LLVM for target (and patch it accordingly, using the patches we use for main LLVM) and Firefox now builds successfully. Therefore, I can confirm that this is probably a bug in LLVM7, and Firefox triggers it somehow in that one crate. I'd say switching is the most reasonable thing to do, as it gives us no disadvantages (besides a bit longer build) but also gives us LLVM patched and tested for Rust.

**Comment by jnbr** [comment]
Arch reported a similar issue upstream: https://github.com/rust-lang/rust/issues/57762

**Comment by q66** [comment]
Firefox also uses lto. Disabling it would probably have a significant effect on performance though. I think using bundled LLVM is the right thing to do here; I will test whether cross compiling works and then we should be good to go.

**Comment by q66** [comment]
Here is a full diff needed to make Rust with bundled LLVM build, both native and cross: https://github.com/void-ppc64/void-packages/commit/4b159700aefea7d5dd4776a5cdd3fef24cec4534

In cross, it is after all not possible (or at least not easily) to use host's LLVM, as the target requires `llvm-tblgen` as well as `llvm-config` for the native architecture in version and configuration matching the bundled LLVM. It is possible to build `llvm-tblgen` relatively easily, same with `llvm-config`, but if the cmake configuration is not the same, `llvm-config` will provide incorrect results, so just build LLVM for both host and target.

Even when building two LLVMs, it fortunately only uses the core, so it doesn't add that much to the build time. Normal cross build (using system LLVM) takes about an hour here, with bundled LLVM about 1:15.

**Comment by Cogitri** [comment]
Thank you very much, @q66! Updated the PR, will do some testing now :)

**Comment by jnbr** [comment]
upstream fix for llvm 7: https://github.com/rust-lang/rust/commit/1c95f5a34c14f08d65cdd198827e3a2fcb63cf39

**Comment by q66** [comment]
Looking good. Do we want to get it downstream or wait for 1.32.1?

**Comment by jnbr** [comment]
I would rather get it downstream. There is no sign of a 1.32.1 release yet.

**Comment by Cogitri** [comment]
Alright, applied the patch, optionalised internal-llvm in case we have to do use internal-llvm at some point.

File: srcpkgs/rust/template [hunk 2]
```diff
@@ -58,23 +62,70 @@ else
...
+			patch -sNp1 -i ${x}
+		done
+		popd
+	else
+		# patches for system LLVM
+		for x in ${FILESDIR}/patches/sys-llvm/*.patch; do
+			msg_normal "Applying patch $x\n"
+			patch -sNp1 -i ${x}
+		done
+	fi
```
Comment (Cogitri) [inline srcpkgs/rust/template]: I guess this is a little overkill, but oh well

File: srcpkgs/rust/template (hunk 1 above)
Comment (Cogitri) [reply srcpkgs/rust/template]: So....@Gottox?

File: srcpkgs/rust/template [hunk 3]
```diff
@@ -58,23 +62,70 @@ else
...
+	ppc64le) _build_triplet=powerpc64le-unknown-linux-gnu;;
+	ppc64le-musl) _build_triplet=powerpc64le-unknown-linux-musl;;
+	ppc64-musl) _build_triplet=powerpc64-unknown-linux-musl;;
 	*-musl) _build_triplet=${XBPS_MACHINE%-musl}-unknown-linux-musl;;
 	*) _build_triplet=${XBPS_MACHINE}-unknown-linux-gnu;;
 esac
 
 post_extract() {
-	rm -rf src/llvm
+if [ "$build_option_internal_llvm" ]; then
```
Comment (jnbr) [inline srcpkgs/rust/template]: wrong indentation

**Comment by jnbr** [comment]
> Alright, applied the patch, optionalised internal-llvm in case we have to do use internal-llvm at some point.

looks like a lot of work for a eventuality that should never happen

**Comment by q66** [comment]
the `LDFLAGS=""` workaround has nothing to do with LLVM version used, it works around a completely different issue, and I'm still not convinced it will work applied that way, it should go in the same place where the `CFLAGS` overrides are applied (i.e. as a part of the `env` command).

The Rust patch is for Rust itself, so it belongs in `patches/`, just like patches for any other template, and should get applied automatically.

**Comment by Cogitri** [comment]
> The LDFLAGS="" workaround has nothing to do with LLVM version used, it works around a completely different issue

Ah right, it was for lzma-sys and not for LLVM, my bad.

> The Rust patch is for Rust itself

Sure, but only is required when using LLVM <8, so it's not required for internal LLVM. Doesn't hurt having it in patches anyway.

> looks like a lot of work for a eventuality that should never happen

Hum, doesn't seem like a lot of work to me and as we've seen here it doesn't seem like _such_ a bad idea to have the option to build with internal-llvm.

File: srcpkgs/rust/template [hunk 4]
```diff
@@ -107,14 +159,15 @@ do_configure() {
...
 
-	# Disable jemalloc for now. It increases the size of small programs
-	# significantly (hello world without jemalloc 130KB vs with jemalloc
-	# 300KB) and provides worse performance in some cases.
-	configure_args+=" --disable-jemalloc"
+	if ! [ "$build_option_internal_llvm" ]; then
+		configure_args+="
+			--llvm-root=/usr
+			--set=target.${_build_triplet}.llvm-config=/usr/bin/llvm-config
+			--set=target.${RUST_TARGET}.llvm-config="/usr/bin/llvm-config"
```
Comment (jnbr) [inline srcpkgs/rust/template]: nested quoting, might work but looks broken

**Comment by jnbr** [comment]
1c95f5a34c14f08d65cdd198827e3a2fcb63cf39.patch is redundant

**Comment by Cogitri** [comment]
> 1c95f5a34c14f08d65cdd198827e3a2fcb63cf39.patch is redundant

Eh?

**Comment by jnbr** [comment]
there is
srcpkgs/rust/files/patches/sys-llvm/1c95f5a34c14f08d65cdd198827e3a2fcb63cf39.patch
and
srcpkgs/rust/patches/1c95f5a34c14f08d65cdd198827e3a2fcb63cf39.patch

**Comment by q66** [comment]
> Hum, I guess that requires downstream patching?

As far as I can see, all it takes is passing `--llvm-link-shared` to `configure`

**Comment by Cogitri** [comment]
> there is
srcpkgs/rust/files/patches/sys-llvm/1c95f5a34c14f08d65cdd198827e3a2fcb63cf39.patch
and
srcpkgs/rust/patches/1c95f5a34c14f08d65cdd198827e3a2fcb63cf39.patch

Oops, deleted it but forgot to git add, so I didn't see it anymore

> As far as I can see, all it takes is passing --llvm-link-shared to configure

Alright, let's try that

**Comment by Cogitri** [comment]
> configure:
configure: error: Option '--llvm-link-shared' is not recognized

Let's start digging

**Comment by Cogitri** [comment]
Oops, it's `--enable-llvm-link-shared`

**Comment by jnbr** [comment]
`CARGO_BUILD_TARGET` (added to global env in 027cdb5bb8153102ef77ab595a07a44d7dc1f2e9) needs to be unset to fix cross builds

**Comment by q66** [comment]
Just tested `ppc64le` native, worked fine. Cross builds fail, probably because of what @jnbr said above. I can confirm that `llvm-with-dependencies.patch` can be dropped. Also, the comment about unsetting `LDFLAGS` is in a wrong place.

**Comment by q66** [comment]
I successfully cross-compiled rust to `ppc64le-musl` and `ppc64-musl`. Here is my patch: https://github.com/void-ppc64/void-packages/commit/67e76c5dd1bfb0a75fe525f9c9c7ec55978b3e11

It's mostly the same as in this PR except

1) it unsets `CARGO_BUILD_TARGET`
2) moves the comment to the right place
3) drops `llvm-with-dependencies.patch` as it's unnecessary with shared `llvm`

**Comment by Cogitri** [comment]
Updated.

**Comment by q66** [comment]
Looks good to me

File: srcpkgs/rust/template (hunk 1 above)
Comment (jnbr) [reply srcpkgs/rust/template]: @Cogitri files are on /distfiles, can you change the URLs?

**Comment by Cogitri** [comment]
Update distfiles URL

**Comment by jnbr** [comment]
merged in 4aa64287db41751191923267bf02f9d9c0ace3fe

**Pull request closed without merging**